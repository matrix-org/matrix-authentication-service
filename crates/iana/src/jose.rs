// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Enums from the "JSON Object Signing and Encryption" IANA registry
//! See <https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml>

// Do not edit this file manually

use parse_display::{Display, FromStr};
use schemars::JsonSchema;
use serde_with::{DeserializeFromStr, SerializeDisplay};

/// JSON Web Signature "alg" parameter
///
/// Source: <https://www.iana.org/assignments/jose/web-signature-encryption-algorithms.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebSignatureAlg {
    /// HMAC using SHA-256
    #[schemars(rename = "HS256")]
    #[display("HS256")]
    Hs256,

    /// HMAC using SHA-384
    #[schemars(rename = "HS384")]
    #[display("HS384")]
    Hs384,

    /// HMAC using SHA-512
    #[schemars(rename = "HS512")]
    #[display("HS512")]
    Hs512,

    /// RSASSA-PKCS1-v1_5 using SHA-256
    #[schemars(rename = "RS256")]
    #[display("RS256")]
    Rs256,

    /// RSASSA-PKCS1-v1_5 using SHA-384
    #[schemars(rename = "RS384")]
    #[display("RS384")]
    Rs384,

    /// RSASSA-PKCS1-v1_5 using SHA-512
    #[schemars(rename = "RS512")]
    #[display("RS512")]
    Rs512,

    /// ECDSA using P-256 and SHA-256
    #[schemars(rename = "ES256")]
    #[display("ES256")]
    Es256,

    /// ECDSA using P-384 and SHA-384
    #[schemars(rename = "ES384")]
    #[display("ES384")]
    Es384,

    /// ECDSA using P-521 and SHA-512
    #[schemars(rename = "ES512")]
    #[display("ES512")]
    Es512,

    /// RSASSA-PSS using SHA-256 and MGF1 with SHA-256
    #[schemars(rename = "PS256")]
    #[display("PS256")]
    Ps256,

    /// RSASSA-PSS using SHA-384 and MGF1 with SHA-384
    #[schemars(rename = "PS384")]
    #[display("PS384")]
    Ps384,

    /// RSASSA-PSS using SHA-512 and MGF1 with SHA-512
    #[schemars(rename = "PS512")]
    #[display("PS512")]
    Ps512,

    /// No digital signature or MAC performed
    #[schemars(rename = "none")]
    #[display("none")]
    None,

    /// EdDSA signature algorithms
    #[schemars(rename = "EdDSA")]
    #[display("EdDSA")]
    EdDsa,

    /// ECDSA using secp256k1 curve and SHA-256
    #[schemars(rename = "ES256K")]
    #[display("ES256K")]
    Es256K,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Encryption "alg" parameter
///
/// Source: <https://www.iana.org/assignments/jose/web-signature-encryption-algorithms.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebEncryptionAlg {
    /// RSAES-PKCS1-v1_5
    #[schemars(rename = "RSA1_5")]
    #[display("RSA1_5")]
    Rsa15,

    /// RSAES OAEP using default parameters
    #[schemars(rename = "RSA-OAEP")]
    #[display("RSA-OAEP")]
    RsaOaep,

    /// RSAES OAEP using SHA-256 and MGF1 with SHA-256
    #[schemars(rename = "RSA-OAEP-256")]
    #[display("RSA-OAEP-256")]
    RsaOaep256,

    /// AES Key Wrap using 128-bit key
    #[schemars(rename = "A128KW")]
    #[display("A128KW")]
    A128Kw,

    /// AES Key Wrap using 192-bit key
    #[schemars(rename = "A192KW")]
    #[display("A192KW")]
    A192Kw,

    /// AES Key Wrap using 256-bit key
    #[schemars(rename = "A256KW")]
    #[display("A256KW")]
    A256Kw,

    /// Direct use of a shared symmetric key
    #[schemars(rename = "dir")]
    #[display("dir")]
    Dir,

    /// ECDH-ES using Concat KDF
    #[schemars(rename = "ECDH-ES")]
    #[display("ECDH-ES")]
    EcdhEs,

    /// ECDH-ES using Concat KDF and "A128KW" wrapping
    #[schemars(rename = "ECDH-ES+A128KW")]
    #[display("ECDH-ES+A128KW")]
    EcdhEsA128Kw,

    /// ECDH-ES using Concat KDF and "A192KW" wrapping
    #[schemars(rename = "ECDH-ES+A192KW")]
    #[display("ECDH-ES+A192KW")]
    EcdhEsA192Kw,

    /// ECDH-ES using Concat KDF and "A256KW" wrapping
    #[schemars(rename = "ECDH-ES+A256KW")]
    #[display("ECDH-ES+A256KW")]
    EcdhEsA256Kw,

    /// Key wrapping with AES GCM using 128-bit key
    #[schemars(rename = "A128GCMKW")]
    #[display("A128GCMKW")]
    A128Gcmkw,

    /// Key wrapping with AES GCM using 192-bit key
    #[schemars(rename = "A192GCMKW")]
    #[display("A192GCMKW")]
    A192Gcmkw,

    /// Key wrapping with AES GCM using 256-bit key
    #[schemars(rename = "A256GCMKW")]
    #[display("A256GCMKW")]
    A256Gcmkw,

    /// PBES2 with HMAC SHA-256 and "A128KW" wrapping
    #[schemars(rename = "PBES2-HS256+A128KW")]
    #[display("PBES2-HS256+A128KW")]
    Pbes2Hs256A128Kw,

    /// PBES2 with HMAC SHA-384 and "A192KW" wrapping
    #[schemars(rename = "PBES2-HS384+A192KW")]
    #[display("PBES2-HS384+A192KW")]
    Pbes2Hs384A192Kw,

    /// PBES2 with HMAC SHA-512 and "A256KW" wrapping
    #[schemars(rename = "PBES2-HS512+A256KW")]
    #[display("PBES2-HS512+A256KW")]
    Pbes2Hs512A256Kw,

    /// RSA-OAEP using SHA-384 and MGF1 with SHA-384
    #[schemars(rename = "RSA-OAEP-384")]
    #[display("RSA-OAEP-384")]
    RsaOaep384,

    /// RSA-OAEP using SHA-512 and MGF1 with SHA-512
    #[schemars(rename = "RSA-OAEP-512")]
    #[display("RSA-OAEP-512")]
    RsaOaep512,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Encryption "enc" parameter
///
/// Source: <https://www.iana.org/assignments/jose/web-signature-encryption-algorithms.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebEncryptionEnc {
    /// AES_128_CBC_HMAC_SHA_256 authenticated encryption algorithm
    #[schemars(rename = "A128CBC-HS256")]
    #[display("A128CBC-HS256")]
    A128CbcHs256,

    /// AES_192_CBC_HMAC_SHA_384 authenticated encryption algorithm
    #[schemars(rename = "A192CBC-HS384")]
    #[display("A192CBC-HS384")]
    A192CbcHs384,

    /// AES_256_CBC_HMAC_SHA_512 authenticated encryption algorithm
    #[schemars(rename = "A256CBC-HS512")]
    #[display("A256CBC-HS512")]
    A256CbcHs512,

    /// AES GCM using 128-bit key
    #[schemars(rename = "A128GCM")]
    #[display("A128GCM")]
    A128Gcm,

    /// AES GCM using 192-bit key
    #[schemars(rename = "A192GCM")]
    #[display("A192GCM")]
    A192Gcm,

    /// AES GCM using 256-bit key
    #[schemars(rename = "A256GCM")]
    #[display("A256GCM")]
    A256Gcm,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Encryption Compression Algorithm
///
/// Source: <https://www.iana.org/assignments/jose/web-encryption-compression-algorithms.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebEncryptionCompressionAlgorithm {
    /// DEFLATE
    #[schemars(rename = "DEF")]
    #[display("DEF")]
    Def,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Key Type
///
/// Source: <https://www.iana.org/assignments/jose/web-key-types.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebKeyType {
    /// Elliptic Curve
    #[schemars(rename = "EC")]
    #[display("EC")]
    Ec,

    /// RSA
    #[schemars(rename = "RSA")]
    #[display("RSA")]
    Rsa,

    /// Octet sequence
    #[schemars(rename = "oct")]
    #[display("oct")]
    Oct,

    /// Octet string key pairs
    #[schemars(rename = "OKP")]
    #[display("OKP")]
    Okp,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Key EC Elliptic Curve
///
/// Source: <https://www.iana.org/assignments/jose/web-key-elliptic-curve.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebKeyEcEllipticCurve {
    /// P-256 Curve
    #[schemars(rename = "P-256")]
    #[display("P-256")]
    P256,

    /// P-384 Curve
    #[schemars(rename = "P-384")]
    #[display("P-384")]
    P384,

    /// P-521 Curve
    #[schemars(rename = "P-521")]
    #[display("P-521")]
    P521,

    /// SECG secp256k1 curve
    #[schemars(rename = "secp256k1")]
    #[display("secp256k1")]
    Secp256K1,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Key OKP Elliptic Curve
///
/// Source: <https://www.iana.org/assignments/jose/web-key-elliptic-curve.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebKeyOkpEllipticCurve {
    /// Ed25519 signature algorithm key pairs
    #[schemars(rename = "Ed25519")]
    #[display("Ed25519")]
    Ed25519,

    /// Ed448 signature algorithm key pairs
    #[schemars(rename = "Ed448")]
    #[display("Ed448")]
    Ed448,

    /// X25519 function key pairs
    #[schemars(rename = "X25519")]
    #[display("X25519")]
    X25519,

    /// X448 function key pairs
    #[schemars(rename = "X448")]
    #[display("X448")]
    X448,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Key Use
///
/// Source: <https://www.iana.org/assignments/jose/web-key-use.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebKeyUse {
    /// Digital Signature or MAC
    #[schemars(rename = "sig")]
    #[display("sig")]
    Sig,

    /// Encryption
    #[schemars(rename = "enc")]
    #[display("enc")]
    Enc,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}

/// JSON Web Key Operation
///
/// Source: <https://www.iana.org/assignments/jose/web-key-operations.csv>
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Display,
    FromStr,
    SerializeDisplay,
    DeserializeFromStr,
    JsonSchema,
)]
#[non_exhaustive]
pub enum JsonWebKeyOperation {
    /// Compute digital signature or MAC
    #[schemars(rename = "sign")]
    #[display("sign")]
    Sign,

    /// Verify digital signature or MAC
    #[schemars(rename = "verify")]
    #[display("verify")]
    Verify,

    /// Encrypt content
    #[schemars(rename = "encrypt")]
    #[display("encrypt")]
    Encrypt,

    /// Decrypt content and validate decryption, if applicable
    #[schemars(rename = "decrypt")]
    #[display("decrypt")]
    Decrypt,

    /// Encrypt key
    #[schemars(rename = "wrapKey")]
    #[display("wrapKey")]
    WrapKey,

    /// Decrypt key and validate decryption, if applicable
    #[schemars(rename = "unwrapKey")]
    #[display("unwrapKey")]
    UnwrapKey,

    /// Derive key
    #[schemars(rename = "deriveKey")]
    #[display("deriveKey")]
    DeriveKey,

    /// Derive bits not to be used as a key
    #[schemars(rename = "deriveBits")]
    #[display("deriveBits")]
    DeriveBits,

    /// An unknown value.
    #[display("{0}")]
    #[schemars(skip)]
    Unknown(String),
}
