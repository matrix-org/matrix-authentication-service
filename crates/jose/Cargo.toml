[package]
name = "mas-jose"
version = "0.1.0"
authors = ["Quentin Gliech <quenting@element.io>"]
edition = "2021"
license = "Apache-2.0"

[dependencies]
base64ct = { version = "1.6.0", features = ["std"] }
chrono = { version = "0.4.24", features = ["serde"] }
digest = "0.10.6"
ecdsa = { version = "0.16.1", features = ["signing", "verifying"] }
elliptic-curve = "0.13.2"
generic-array = "0.14.6"
hmac = "0.12.1"
k256 = { version = "0.13.0", features = ["ecdsa"] }
p256 = { version = "0.13.0", features = ["ecdsa"] }
p384 = { version = "0.13.0", features = ["ecdsa"] }
rand = "0.8.5"
rsa = "0.9.0-pre.0"
schemars = "0.8.12"
sec1 = "0.7.1"
serde = { version = "1.0.158", features = ["derive"] }
serde_json = "1.0.94"
serde_with = { version = "2.3.1", features = ["base64"] }
sha2 = { version = "0.10.6", features = ["oid"] }
signature = "2.0.0"
thiserror = "1.0.40"
tracing = "0.1.37"
url = { version = "2.3.1", features = ["serde"] }

mas-iana = { path = "../iana" }

[dev-dependencies]
insta = { version = "1.28.0" }
rand_chacha = "0.3.1"
