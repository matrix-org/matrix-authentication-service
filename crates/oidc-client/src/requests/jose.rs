// Copyright 2022 KÃ©vin Commaille.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Requests and method related to JSON Object Signing and Encryption.

use std::collections::HashMap;

use bytes::Bytes;
use chrono::{DateTime, Utc};
use mas_http::JsonResponseLayer;
use mas_iana::jose::JsonWebSignatureAlg;
use mas_jose::{
    claims::{self, TimeOptions},
    jwk::PublicJsonWebKeySet,
    jwt::Jwt,
};
use serde_json::Value;
use tower::{Layer, Service, ServiceExt};
use url::Url;

use crate::{
    error::{IdTokenError, JwksError, JwtVerificationError},
    http_service::HttpService,
    types::IdToken,
};

/// Fetch a JWKS at the given URL.
///
/// # Arguments
///
/// * `http_service` - The service to use for making HTTP requests.
///
/// * `jwks_uri` - The URL where the JWKS can be retrieved.
///
/// # Errors
///
/// Returns an error if the request fails or if the data is invalid.
#[tracing::instrument(skip_all, fields(jwks_uri))]
pub async fn fetch_jwks(
    http_service: &HttpService,
    jwks_uri: &Url,
) -> Result<PublicJsonWebKeySet, JwksError> {
    tracing::debug!("Fetching JWKS...");

    let jwks_request = http::Request::get(jwks_uri.as_str()).body(Bytes::new())?;

    let service = JsonResponseLayer::<PublicJsonWebKeySet>::default().layer(http_service.clone());

    let response = service.ready_oneshot().await?.call(jwks_request).await?;

    Ok(response.into_body())
}

/// The data required to verify a JWT.
#[derive(Clone, Copy)]
pub struct JwtVerificationData<'a> {
    /// The URL of the issuer that generated the ID Token.
    pub issuer: &'a Url,

    /// The issuer's JWKS.
    pub jwks: &'a PublicJsonWebKeySet,

    /// The ID obtained when registering the client.
    pub client_id: &'a String,

    /// The JWA that should have been used to sign the JWT, as set during
    /// client registration.
    pub signing_algorithm: &'a JsonWebSignatureAlg,
}

/// Decode and verify a signed JWT.
///
/// The following checks are performed:
///
/// * The signature is verified with the given JWKS.
///
/// * The `iss` claim must be present and match the issuer.
///
/// * The `aud` claim must be present and match the client ID.
///
/// * The `alg` in the header must match the signing algorithm.
///
/// # Arguments
///
/// * `jwt` - The serialized JWT to decode and verify.
///
/// * `jwks` - The JWKS that should contain the public key to verify the JWT's
///   signature.
///
/// * `issuer` - The issuer of the JWT.
///
/// * `audience` - The audience that the JWT is intended for.
///
/// * `signing_algorithm` - The JWA that should have been used to sign the JWT.
///
/// # Errors
///
/// Returns an error if the data is invalid or verification fails.
pub fn verify_signed_jwt<'a>(
    jwt: &'a str,
    verification_data: JwtVerificationData<'_>,
) -> Result<Jwt<'a, HashMap<String, Value>>, JwtVerificationError> {
    tracing::debug!("Validating JWT...");

    let JwtVerificationData {
        issuer,
        jwks,
        client_id,
        signing_algorithm,
    } = verification_data;

    let jwt: Jwt<HashMap<String, Value>> = jwt.try_into()?;

    jwt.verify_with_jwks(jwks)?;

    let (header, mut claims) = jwt.clone().into_parts();

    // Must have the proper issuer.
    claims::ISS.extract_required_with_options(&mut claims, issuer.as_str())?;

    // Must have the proper audience.
    claims::AUD.extract_required_with_options(&mut claims, client_id)?;

    // Must use the proper algorithm.
    if header.alg() != signing_algorithm {
        return Err(JwtVerificationError::WrongSignatureAlg);
    }

    Ok(jwt)
}

/// Decode and verify an ID Token.
///
/// Besides the checks of [`verify_signed_jwt()`], the following checks are
/// performed:
///
/// * The `exp` claim must be present and the token must not have expired.
///
/// * The `iat` claim must be present must be in the past.
///
/// * The `sub` claim must be present.
///
/// If an authorization ID token is provided, these extra checks are performed:
///
/// * The `sub` claims must match.
///
/// * The `auth_time` claims must match.
///
/// # Arguments
///
/// * `id_token` - The serialized ID Token to decode and verify.
///
/// * `verification_data` - The data necessary to verify the ID Token.
///
/// * `auth_id_token` - If the ID Token is not verified during an authorization
///   request, the ID token that was returned from the latest authorization
///   request.
///
/// # Errors
///
/// Returns an error if the data is invalid or verification fails.
pub fn verify_id_token<'a>(
    id_token: &'a str,
    verification_data: JwtVerificationData<'_>,
    auth_id_token: Option<&IdToken<'_>>,
    now: DateTime<Utc>,
) -> Result<IdToken<'a>, IdTokenError> {
    let id_token = verify_signed_jwt(id_token, verification_data)?;

    let mut claims = id_token.payload().clone();

    let time_options = TimeOptions::new(now);
    // Must not have expired.
    claims::EXP.extract_required_with_options(&mut claims, &time_options)?;

    // `iat` claim must be present.
    claims::IAT.extract_required_with_options(&mut claims, time_options)?;

    // Subject identifier must be present.
    let sub = claims::SUB.extract_required(&mut claims)?;

    // No more checks if there is no previous ID token.
    let auth_id_token = match auth_id_token {
        Some(id_token) => id_token,
        None => return Ok(id_token),
    };

    let mut auth_claims = auth_id_token.payload().clone();

    // Subject identifier must always be the same.
    let auth_sub = claims::SUB.extract_required(&mut auth_claims)?;
    if sub != auth_sub {
        return Err(IdTokenError::WrongSubjectIdentifier);
    }

    // If the authentication time is present, it must be unchanged.
    if let Some(auth_time) = claims::AUTH_TIME.extract_optional(&mut claims)? {
        let prev_auth_time = claims::AUTH_TIME.extract_required(&mut auth_claims)?;

        if prev_auth_time != auth_time {
            return Err(IdTokenError::WrongAuthTime);
        }
    }

    Ok(id_token)
}
