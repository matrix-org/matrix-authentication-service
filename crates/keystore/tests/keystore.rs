// Copyright 2022 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use der::pem::LineEnding;
use mas_iana::jose::JsonWebSignatureAlg;
use mas_jose::{
    jwk::ParametersInfo,
    jwt::{JsonWebSignatureHeader, Jwt},
};
use mas_keystore::{JsonWebKey, JsonWebKeySet, Keystore, PrivateKey};
use rand::SeedableRng;

static PASSWORD: &str = "hunter2";

/// Generate a test which loads a key, and then tries signing and verifying a
/// JWT for each available algorithm
macro_rules! plain_test {
    ($name:ident, $kind:ident, $path:literal) => {
        #[test]
        fn $name() {
            let bytes = include_bytes!(concat!("./keys/", $path));
            let key = PrivateKey::load(bytes).unwrap();
            assert!(matches!(key, PrivateKey::$kind(_)), "wrong key type");

            let algs = key.possible_algs();
            assert_ne!(algs.len(), 0);

            for alg in algs {
                let header = JsonWebSignatureHeader::new(alg.clone());
                let payload = "hello";
                let signer = key.signing_key_for_alg(alg).unwrap();
                let jwt = Jwt::sign(header, payload, &signer).unwrap();
                let verifier = key.verifying_key_for_alg(alg).unwrap();
                jwt.verify(&verifier).unwrap();
            }
        }
    };
}

/// Generate a test which loads an encrypted key, and then tries signing and
/// verifying a JWT for each available algorithm
macro_rules! enc_test {
    ($name:ident, $kind:ident, $path:literal) => {
        #[test]
        fn $name() {
            let bytes = include_bytes!(concat!("./keys/", $path));
            let key = PrivateKey::load_encrypted(bytes, PASSWORD).unwrap();
            assert!(matches!(key, PrivateKey::$kind(_)), "wrong key type");

            let algs = key.possible_algs();
            assert_ne!(algs.len(), 0);

            for alg in algs {
                let header = JsonWebSignatureHeader::new(alg.clone());
                let payload = "hello";
                let signer = key.signing_key_for_alg(alg).unwrap();
                let jwt = Jwt::sign(header, payload, &signer).unwrap();
                let verifier = key.verifying_key_for_alg(alg).unwrap();
                jwt.verify(&verifier).unwrap();
            }
        }
    };
}

/// Generate a PEM decoding and encoding test
macro_rules! pem_test {
    ($name:ident, $path:literal) => {
        pem_test!($name, $path, compare = true);
    };
    ($name:ident, $path:literal, compare = $compare:literal) => {
        #[test]
        fn $name() {
            let pem = include_str!(concat!("./keys/", $path, ".pem"));
            let key = PrivateKey::load_pem(pem).unwrap();
            let pem2 = key.to_pem(pem_rfc7468::LineEnding::LF).unwrap();

            if $compare {
                assert_eq!(pem, pem2.as_str());
            }
        }
    };
}

/// Generate a DER decoding and encoding test
macro_rules! der_test {
    ($name:ident, $path:literal) => {
        der_test!($name, $path, compare = true);
    };
    ($name:ident, $path:literal, compare = $compare:literal) => {
        #[test]
        fn $name() {
            let der = include_bytes!(concat!("./keys/", $path, ".der"));
            let key = PrivateKey::load_der(der).unwrap();
            let der2 = key.to_der().unwrap();

            if $compare {
                assert_eq!(der, der2.as_slice());
            }
        }
    };
}

plain_test!(plain_rsa_pkcs1_pem, Rsa, "rsa.pkcs1.pem");
plain_test!(plain_rsa_pkcs1_der, Rsa, "rsa.pkcs1.der");
plain_test!(plain_rsa_pkcs8_pem, Rsa, "rsa.pkcs8.pem");
plain_test!(plain_rsa_pkcs8_der, Rsa, "rsa.pkcs8.der");
plain_test!(plain_ec_p256_sec1_pem, EcP256, "ec-p256.sec1.pem");
plain_test!(plain_ec_p256_sec1_der, EcP256, "ec-p256.sec1.der");
plain_test!(plain_ec_p256_pkcs8_pem, EcP256, "ec-p256.pkcs8.pem");
plain_test!(plain_ec_p256_pkcs8_der, EcP256, "ec-p256.pkcs8.der");
plain_test!(plain_ec_p384_sec1_pem, EcP384, "ec-p384.sec1.pem");
plain_test!(plain_ec_p384_sec1_der, EcP384, "ec-p384.sec1.der");
plain_test!(plain_ec_p384_pkcs8_pem, EcP384, "ec-p384.pkcs8.pem");
plain_test!(plain_ec_p384_pkcs8_der, EcP384, "ec-p384.pkcs8.der");
plain_test!(plain_ec_k256_sec1_pem, EcK256, "ec-k256.sec1.pem");
plain_test!(plain_ec_k256_sec1_der, EcK256, "ec-k256.sec1.der");
plain_test!(plain_ec_k256_pkcs8_pem, EcK256, "ec-k256.pkcs8.pem");
plain_test!(plain_ec_k256_pkcs8_der, EcK256, "ec-k256.pkcs8.der");

enc_test!(enc_rsa_pkcs8_pem, Rsa, "rsa.pkcs8.encrypted.pem");
enc_test!(enc_rsa_pkcs8_der, Rsa, "rsa.pkcs8.encrypted.der");
enc_test!(enc_ec_p256_pkcs8_pem, EcP256, "ec-p256.pkcs8.encrypted.pem");
enc_test!(enc_ec_p256_pkcs8_der, EcP256, "ec-p256.pkcs8.encrypted.der");
enc_test!(enc_ec_p384_pkcs8_pem, EcP384, "ec-p384.pkcs8.encrypted.pem");
enc_test!(enc_ec_p384_pkcs8_der, EcP384, "ec-p384.pkcs8.encrypted.der");
enc_test!(enc_ec_k256_pkcs8_pem, EcK256, "ec-k256.pkcs8.encrypted.pem");
enc_test!(enc_ec_k256_pkcs8_der, EcK256, "ec-k256.pkcs8.encrypted.der");

// Test PEM/DER serialization. For some reason, DER serialization of EC keys
// have some extra bytes
pem_test!(serialize_rsa_pkcs1_pem, "rsa.pkcs1");
der_test!(serialize_rsa_pkcs1_der, "rsa.pkcs1");
pem_test!(serialize_ec_p256_sec1_pem, "ec-p256.sec1");
der_test!(serialize_ec_p256_sec1_der, "ec-p256.sec1", compare = false);
pem_test!(serialize_ec_p384_sec1_pem, "ec-p384.sec1");
der_test!(serialize_ec_p384_sec1_der, "ec-p384.sec1", compare = false);
pem_test!(serialize_ec_k256_sec1_pem, "ec-k256.sec1");
der_test!(serialize_ec_k256_sec1_der, "ec-k256.sec1", compare = false);

#[test]
fn load_encrypted_as_unencrypted_error() {
    let pem = include_str!("./keys/rsa.pkcs8.encrypted.pem");
    assert!(PrivateKey::load_pem(pem).unwrap_err().is_encrypted());

    let der = include_bytes!("./keys/rsa.pkcs8.encrypted.der");
    assert!(PrivateKey::load_der(der).unwrap_err().is_encrypted());
}

#[test]
fn load_unencrypted_as_encrypted_error() {
    let pem = include_str!("./keys/rsa.pkcs8.pem");
    assert!(PrivateKey::load_encrypted_pem(pem, PASSWORD)
        .unwrap_err()
        .is_unencrypted());

    let der = include_bytes!("./keys/rsa.pkcs8.der");
    assert!(PrivateKey::load_encrypted_der(der, PASSWORD)
        .unwrap_err()
        .is_unencrypted());
}

#[test]
fn generate_sign_and_verify() {
    // Use a seeded RNG to keep the snapshot stable
    let mut rng = rand_chacha::ChaCha8Rng::seed_from_u64(42);

    let rsa = PrivateKey::generate_rsa(&mut rng).expect("Failed to generate RSA key");
    insta::assert_snapshot!(rsa.to_pem(LineEnding::LF).unwrap());

    let ec_p256 = PrivateKey::generate_ec_p256(&mut rng);
    insta::assert_snapshot!(ec_p256.to_pem(LineEnding::LF).unwrap());

    let ec_p384 = PrivateKey::generate_ec_p384(&mut rng);
    insta::assert_snapshot!(ec_p384.to_pem(LineEnding::LF).unwrap());

    let ec_k256 = PrivateKey::generate_ec_k256(&mut rng);
    insta::assert_snapshot!(ec_k256.to_pem(LineEnding::LF).unwrap());

    // Create a keystore out of the keys
    let keyset = Keystore::new(JsonWebKeySet::new(vec![
        JsonWebKey::new(rsa),
        JsonWebKey::new(ec_p256),
        JsonWebKey::new(ec_p384),
        JsonWebKey::new(ec_k256),
    ]));

    // And extract the public JWKS
    let jwks = keyset.public_jwks();
    insta::assert_yaml_snapshot!(jwks);

    // Try signing for each supported algorithm
    for alg in [
        JsonWebSignatureAlg::Rs256,
        JsonWebSignatureAlg::Rs384,
        JsonWebSignatureAlg::Rs512,
        JsonWebSignatureAlg::Ps256,
        JsonWebSignatureAlg::Ps384,
        JsonWebSignatureAlg::Ps512,
        JsonWebSignatureAlg::Es256,
        JsonWebSignatureAlg::Es384,
        JsonWebSignatureAlg::Es256K,
    ] {
        // Find a matching key and sign with it
        let key = keyset.signing_key_for_algorithm(&alg).unwrap();
        let signer = key.params().signing_key_for_alg(&alg).unwrap();
        let header = JsonWebSignatureHeader::new(alg.clone());
        let token = Jwt::sign_with_rng(&mut rng, header, "", &signer).unwrap();
        insta::assert_snapshot!(format!("jwt_{alg}"), token.as_str());

        // Then try to verify from the public JWKS
        token.verify_with_jwks(&jwks).unwrap();
    }
}
