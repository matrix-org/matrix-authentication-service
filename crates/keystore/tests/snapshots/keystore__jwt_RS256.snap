---
source: crates/keystore/tests/keystore.rs
expression: token.as_str()
---
eyJhbGciOiJSUzI1NiJ9.IiI.H_iz_ry505dPMNxQIi4raU88i3wu7DS0Rre6Qm_LJz0Ee_gd5C_t92fBcrFkber1XL_p8AvlXx4DT2Zr_PMBL_2IblJ7t0Od5wnGC__twarj0v1t6KfUkLXcJ3Jy-StnHNbFTmdFnLuGGWIO7xG9h6xgKIvTroVoLJekMzYCc0wSFiyCfaow4yuKesQHUO-N9VDDPoYhkCPqbhVI_d0y6u7KmQy97FbCdCIxvPGHWrwxWcmYbTh4K9xhGDspDUUEubjYTg3t-oaMc2TJqWvu2FE8jyD02A8OCgca6bCU3NmV_Qr6LSUpFNsL4c-0sIp3-L9ndEWzGnN-ZeeGKur-FQ
